src/core_read.cpp

```diff
@@ -155,3 +156,86 @@ vector<unsigned char> ParseHexUV(const UniValue& v, const string& strName)
         throw runtime_error(strName+" must be hexadecimal string (not '"+strHex+"')");
     return ParseHex(strHex);
 }
+
+static bool ParseKeyTreeNode(const std::string &s, size_t &pos, KeyTreeNode& tree);
+static bool ParseKeyTreeCall(const std::string &s, size_t &pos, unsigned long* num, std::vector<KeyTreeNode>& children)
+{
+    if (s.size() == pos) return false;
+    if (s[pos] != '(') return false;
+    pos++;
+    int count = 0;
+    if (num) {
+        const char *ptr = &s[pos];
+        char *eptr = NULL;
+        *num = strtoul(ptr, &eptr, 10);
+        if (eptr == ptr) return false;
+        pos += eptr - ptr;
+        count++;
+    }
+    while (true) {
+        if (count) {
+            if (pos == s.size()) return false;
+            if (s[pos] == /*(*/')') {
+                pos++;
+                return true;
+            }
+            if (s[pos] != ',') return false;
+            pos++;
+        }
+        children.push_back(KeyTreeNode());
+        if (!ParseKeyTreeNode(s, pos, children.back())) return false;
+        count++;
+    }
+}
+
+static bool ParseKeyTreeNode(const std::string &s, size_t &pos, KeyTreeNode& tree)
+{
+    while (pos < s.size() && isspace(s[pos])) pos++;
+    if (s.size() >= pos + 66 && IsHex(s.substr(pos, 66))) {
+        std::vector<unsigned char> data = ParseHex(s.substr(pos, 66));
+        tree.leaf.Set(data.begin(), data.end());
+        pos += 66;
+        return tree.leaf.IsFullyValid();
+    }
+    if (s.size() >= pos + 2 && s.substr(pos, 2) == "OR") {
+        pos += 2;
+        if (!ParseKeyTreeCall(s, pos, NULL, tree.children)) return false;
+        if (tree.children.size() < 2) return false;
```

Comment:
Using `< 2` rather than `< 1` forbids constructs such as `AND(key)` or `OR(key)` (which would both be synonyms for `key`). I take it this is deliberate?

I think we want better error reporting here, even an integer code.